server/src/services/data_sources.service.ts

```diff
@@ -1,115 +1,245 @@
+      const dataSources = result
+        ?.map((ds) => {
+          if (ds.kind === 'restapi') {
+            const options = {};
+            Object.keys(ds.dataSourceOptions?.[0]?.options || {}).filter((key) => {
+              if (key !== 'tokenData') {
+                return (options[key] = ds.dataSourceOptions[0].options[key]);
+              }
+            });
+            ds.options = options;
+          } else {
+            ds.options = { ...(ds.dataSourceOptions?.[0]?.options || {}) };
+          }
+          delete ds['dataSourceOptions'];
+          return ds;
+        })
+        ?.filter((ds) => ds.kind !== 'restapidefault' && ds.kind !== 'runjsdefault');
 
-    return dataSources;
+      return dataSources;
+    });
   }
 
   async findOne(dataSourceId: string): Promise<DataSource> {
-    return await this.dataSourcesRepository.findOne({
+    return await this.dataSourcesRepository.findOneOrFail({
       where: { id: dataSourceId },
-      relations: ['app', 'plugin'],
+      relations: ['plugin', 'apps', 'dataSourceOptions'],
     });
   }
 
-  async create(
-    name: string,
+  async findOneByEnvironment(dataSourceId: string, environmentId?: string): Promise<DataSource> {
+    const dataSource = await this.dataSourcesRepository.findOneOrFail({
+      where: { id: dataSourceId },
+      relations: ['plugin', 'apps', 'dataSourceOptions'],
+    });
+    if (environmentId) {
+      dataSource.options = await this.appEnvironmentService.getOptions(dataSourceId, null, environmentId);
+    } else {
+      dataSource.options = dataSource.dataSourceOptions?.[0]?.options || {};
+    }
+    return dataSource;
+  }
+
+  async findApp(dataSourceId: string, manager: EntityManager): Promise<App> {
+    return (
+      await manager
+        .createQueryBuilder(DataSource, 'data_source')
+        .innerJoinAndSelect('data_source.apps', 'apps')
+        .where('data_source.id = :dataSourceId', { dataSourceId })
+        .getOneOrFail()
+    ).app;
+  }
+
+  async findDatasourceByKind(kind: string, environmentId: string) {
+    return await dbTransactionWrap(async (manager: EntityManager) => {
+      const currentEnv = environmentId
+        ? await manager.findOneOrFail(AppEnvironment, { where: { id: environmentId } })
+        : await manager.findOneOrFail(AppEnvironment, { where: { isDefault: true } });
+      return await this.dataSourcesRepository.findOneOrFail({
+        where: { kind, appVersionId: currentEnv.appVersionId },
+        relations: ['plugin', 'apps'],
+      });
+    });
+  }
+
+  async findDefaultDataSource(
     kind: string,
-    options: Array<object>,
-    appId: string,
-    appVersionId?: string, // TODO: Make this non optional when autosave is implemented
-    pluginId?: string
+    appVersionId: string,
+    pluginId: string,
+    manager: EntityManager
+  ): Promise<DataSource> {
+    const defaultDataSource = await manager.findOne(DataSource, {
+      where: { kind: `${kind}default`, appVersionId },
```

Comment:
Is it feasible to avoid adding `default` key to the kind? that way can avoid some checks at some places

Reply:
No, since existing restapi data source conflicts. we have to add some privileges or filters for runjs and restapi which is created without a datasource in previous versions. So to differentiate we should add default